### PR TITLE
Set matchers

### PR DESCRIPTION
**Pull request by maksimf**

### CODE REVIEW

File: lib/engine/matchers/part_of_set_matcher.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    def self.matcher_type
+      'PART_OF_SET'.freeze
+    end
+
+    def initialize(remote_array)
+      @remote_set = remote_array.to_set
+    end
+
+    def match?(_key, local_array)
+      local_array.to_set.intersect? @remote_set
```
Comment (patricioe) [inline lib/engine/matchers/part_of_set_matcher.rb]: if the local_array is what the dev passed along in the code, make sure that:

`local_array intersect @remote` is not `true` when `local_array` is empty.

File: spec/engine/matchers/part_of_set_matcher_spec.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  let(:split_json) { File.read(File.expand_path(File.join(File.dirname(__FILE__), '../../test_data/splits/part_of_set_matcher/split.json'))) }
+
+  before do
+    stub_request(:get, 'https://sdk.split.io/api/splitChanges?since=-1')
+      .to_return(status: 200, body: split_json)
+  end
+
+  it 'validates the treatment is ON for correct attribute value' do
+    expect(subject.get_treatment('user', 'LOCAL_IS_PART_OF', %w(a b c))).to eq SplitIoClient::Treatments::ON
```
Comment (patricioe) [inline spec/engine/matchers/part_of_set_matcher_spec.rb]: Do other matchers test are done this way with JSON? doesn't it become more testeable if you test the matcher in isolation? like array (input, operation, output) =>  asesertion?

create a test with empty array, the matcher should be false.

and one with null/nil just to see if you get control

Comment (maksimf) [inline spec/engine/matchers/part_of_set_matcher_spec.rb]: > Do other matchers test are done this way with JSON? doesn't it become more testeable if you test the matcher in isolation? like array (input, operation, output) => asesertion?
Makes sense, I've changed that to test matchers in isolation

**Comment by maksimf** [comment]
@patricioe Ready for review

File: lib/engine/matchers/contains_any_matcher.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    def self.matcher_type
+      'CONTAINS_ANY'.freeze
+    end
+
+    def initialize(attribute, remote_array)
+      super(attribute, remote_array)
+    end
+
+    def match?(_key, data)
+      local_set(data, @attribute).intersect? @remote_set
```
Comment (patricioe) [inline lib/engine/matchers/contains_any_matcher.rb]: what happens is local set is empty?
@SenhorCastor should we expect false?

Comment (patricioe) [inline lib/engine/matchers/contains_any_matcher.rb]: do we need a ` return false if @remote_set.empty?` ?

Comment (maksimf) [inline lib/engine/matchers/contains_any_matcher.rb]: No, here's a test for that: https://github.com/splitio/ruby-client/blob/feature/set_matchers/spec/engine/matchers/contains_any_matcher_spec.rb#L13

**Comment by maksimf** [comment]
@patricioe I've updated this PR with new string matchers, ready to be reviewed/merged

**Pull request merged**